### ISSUE TITLE
Weird freezing with Circle of Blood CD on Linux

### ISSUE THREAD

**Issue by Grieverheart** [kanjitalk755/macemu#177]
On Linux, I mount [https://macintoshgarden.org/games/broken-sword-the-shadow-of-the-templarscircle-of-blood](https://macintoshgarden.org/games/broken-sword-the-shadow-of-the-templarscircle-of-blood) as a `cdrom`. Then copy the installation executable to the desktop, and run the installer on the desktop, after which Basilisk II hangs. This is with the commits of 8th of October. Checked with gdb, and Basilisk II is in an infinite loop running opcode 0x0. Not sure how it gets there. The weird part is that I had a similar issue on Mac. Then I deleted BasiliskII.app and re-downloaded it, and after running it again, it was not hanging anymore...

Note that this is with a Quadra 950 ROM, with a 68040 cpu.

Another issue is that I don't have a physical cdrom on my PC, so /dev/cdrom does not exist. That means I cannot drag and drop a cd rom image because there is no drive already (`drives.empty()`).

**Comment by kanjitalk755** [comment]
> Then I deleted BasiliskII.app and re-downloaded it, and after running it again, it was not hanging anymore...

Nothing can be done without reproducibility.

> Another issue is that I don't have a physical cdrom on my PC, so /dev/cdrom does not exist. That means I cannot drag and drop a cd rom image because there is no drive already (drives.empty()).

Drop mounting is possible even without an optical drive.

**Comment by Grieverheart** [comment]
> Nothing can be done without reproducibility.
Did you try to reproduce on Linux?
> Drop mounting is possible even without an optical drive.
It's not working for me. When it reaches [this line](https://github.com/kanjitalk755/macemu/blob/2533f7ac05912bbde159b61083715c293e3b1f41/BasiliskII/src/cdrom.cpp#L325) `if (!drives.empty()) {` in `CDROMDrop`, the drives are empty. Why does this fail? It tries to add this `PrefsAddString("cdrom", "/dev/cdrom");`, but Basilisk fails to open `/dev/cdrom`.

**Comment by Grieverheart** [comment]
Possibly related, but on Linux, when I expand something on the desktop with Stuffit, the expanded folder does not appear on the desktop until I restart.

**Comment by Grieverheart** [comment]
Another update: On windows it worked fine. I installed it and copied the dsk to Linux. On Linux starting game, but freezes nearly immediately.

**Comment by kanjitalk755** [comment]
> Did you try to reproduce on Linux?

I tried it on Debian 12.
Then, I fell into an infinite loop with pc=0xc0,instruction=0x0208.
The UAE CPU core seems to operate differently under Linux, but it's difficult to debug it.

> It's not working for me.

Working well under Debian.

**Comment by Grieverheart** [comment]
> The UAE CPU core seems to operate differently under Linux, but it's difficult to debug it.

I see, didn't know Basilisk uses the UAE 68k implementation. So then it is a bug in UAE's implementation?
By the way, I moved the same disk image with the installation to a Mac, and the Mac would also freeze when the game starts. Again, I deleted the `BasiliskII.app`, unzipped a fresh one and the problem disappeared. The funny thing is, that I also have a BasiliskII in a different location, which I compiled myself, but this fixes both! What could cause such weird behaviour?

Is this something you think you are going to work on?

> Working well under Debian.

Do you have a physical cd-rom drive on your PC? I had the same issue under Windows. If no cdrom is already mounted, drop mounting does not work on Windows either.

**Comment by kanjitalk755** [comment]
> So then it is a bug in UAE's implementation?

Certainly, the problem isn't necessarily within the UAE core, but it's also difficult to pinpoint.

> What could cause such weird behaviour?

I don't know because I have never encountered such a phenomenon.

> Is this something you think you are going to work on?

I might debug if I find any clues.

> Do you have a physical cd-rom drive on your PC?

The reason why it was okay even without a CDROM drive was because there was the following line in `.basilisk_ii_prefs`.

```
cdrom /dev/poll/cdrom
```

**Comment by ualaa1** [comment]
I somehow got the game to work on Basilisk II.
On the Prefs file, add this line cdrom C:\whereever your cd image is located\Broken Sword.iso
![bandicam 2023-11-23 16-02-57-446](https://github.com/kanjitalk755/macemu/assets/68984176/1808ef7f-a581-4531-97c6-6719998d9a12)

**Comment by rakslice** [comment]
Just the whole vibe of this is weird, like why did not they make a CD installer that would run from read only storage? Why does it throw up a window using Metrowerks I/O stub thing to tell you this like a hello world example? Curious.

**Comment by rakslice** [comment]
The thing in common for the installer failure case between Mac OS X and Linux is execution is getting into the zeros in the same way: The a-trap vector (at `0x28`) is getting stomped to `0x360000`, and then on the next trap (which when the installer is failing to launch is `InitGraf`) we go there and start executing zeroes.

However in Basilisk II on Linux it happens every time you run the installer, and in Basilisk II on Mac OS X it only happens some of the time (with no pattern to the successes and failures or correlation of them to anything I can think to check).

The code fragment doing the stomping looks like:
```
0f97b02c : 4252                CLR.W    (A2)
0f97b02e : 204a                MOVEA.L  A2,A0
0f97b030 : 41e8 0036      LEA      +54(A0),A0
0f97b034 : 2548 000c      MOVE.L   A0,+12(A2) 
0f97b038 : 204a                MOVEA.L  A2,A0
0f97b03a : 41e8 0036      LEA      +54(A0),A0
0f97b03e : 2548 0010      MOVE.L   A0,+16(A2) 
0f97b042 : 204a                MOVEA.L  A2,A0
0f97b044 : 41e8 0036      LEA      +54(A0),A0
0f97b048 : 2548 0026      MOVE.L   A0,+38(A2)  // changes a-trap vector to 0x003699b0
0f97b04c : 204a                MOVEA.L  A2,A0
0f97b04e : 41e8 0036      LEA      +54(A0),A0
0f97b052 : 2548 0014      MOVE.L   A0,+20(A2) 
0f97b056 : 204a                MOVEA.L  A2,A0
0f97b058 : 41e8 0036      LEA      +54(A0),A0
0f97b05c : 2548 002a      MOVE.L   A0,+42(A2)  // changes a-trap vector to 0x00360000
0f97b060 : 204a                MOVEA.L  A2,A0
```
...

On the successful runs a similar code sequence executes from a similar address on the way out of the installer, at which time it does not touch the a-trap vector. So perhaps it is a knock-on effect of whatever is going wrong in the installer rather than the cause.

As an aside:

`0000` isn't an invalid 68k opcode; just straight `00000000` is valid 68k machine code:
```
0000 0000      ORI.B    #$00,D0
```
And it's not a jump so it's not an infinite loop on its own.

But obviously if you jump into some random block of zeroed address space an arbitrary distance from anything you might be there executing in a straight line for a long time.

So when it comes to opcode logging you probably need to get a few consecutive 0 opcodes in before it's of interest, because you're going to see some in the normal system boot-up.

**Comment by rakslice** [comment]
uh...

![Image](https://github.com/user-attachments/assets/7b6fd717-7bbc-4d1b-9389-24a2d03b4bd7)

**Comment by rakslice** [comment]
Ok, so tl;dr: Digital Amusement, who is responsible for this port, also once did a Japanese-only Mac port of Transport Tycoon: https://www.tt-forums.net/viewtopic.php?t=5340

So it's understandable if they left this in by mistake if they were reworking their installer in a hurry.
